core/Napi/ehsm_napi.cpp

```diff
@@ -0,0 +1,150 @@
+#include <string.h>
+#include <stdlib.h>
+#include <cstdint>
+#include "ehsm_provider.h"
+
+using namespace EHsmProvider;
+
+static char* BASE64_ENCHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
+
+static unsigned int Base64Encode(char *pDest, unsigned char *pSrc, unsigned int srclen)
+{
+    unsigned char input[3], output[4];
+    size_t i, index_src = 0, index_dest = 0;
+    for(i = 0; i < srclen; i += 3)
+    {
+        //char [0]
+        input[0] = pSrc[index_src++];
+        output[0] = (char)(input[0] >> 2);
+        pDest[index_dest++] = BASE64_ENCHARS[output[0]];
+
+        //char [1]
+        if(index_src < srclen)
+        {
+            input[1] = pSrc[index_src++];
+            output[1] = (char)(((input[0] & 0x03) << 4) + (input[1] >> 4));
+            pDest[index_dest++] = BASE64_ENCHARS[output[1]];
+        }
+        else
+        {
+            output[1] = (char)((input[0] & 0x03) << 4);
+            pDest[index_dest++] =BASE64_ENCHARS[output[1]];
+            pDest[index_dest++] = '=';
+            pDest[index_dest++] = '=';
+            break;
+        }
+        
+        //char [2]
+        if(index_src < srclen)
+        {
+            input[2] = pSrc[index_src++];
+            output[2] = (char)(((input[1] & 0x0f) << 2) + (input[2] >> 6));
+            pDest[index_dest++] = BASE64_ENCHARS[output[2]];
+        }
+        else
+        {
+            output[2] = (char)((input[1] & 0x0f) << 2);
+            pDest[index_dest++] = BASE64_ENCHARS[output[2]];
+            pDest[index_dest++] = '=';
+            break;
+        }
+
+        //char [3]
+        output[3] = (char)(input[2] & 0x3f);
+        pDest[index_dest++] = BASE64_ENCHARS[output[3]];
+    }
+    //null-terminator
+    pDest[index_dest] = '\n';
+    return index_dest;
+}
+
+extern "C" char* CreateKey_napi(int intMechanism, int intOrigin)
+{
+    EH_RV rv = EHR_FUNCTION_FAILED;
+    EH_KEY_BLOB key_blob;
+    char* rest;
+    int i = 0;
+
+    key_blob.pKeyData = NULL;
+    key_blob.ulKeyLen = 0;
+    
+    rv = Initialize();
+    if (rv != EHR_OK) {
+        printf("Initialize failed 0x%lx\n", rv);
+        return "Initialize failed\n";
+    }
+    printf("Initialize done\n");
+
+
+    if ((rv = CreateKey(intMechanism, (EH_KEY_ORIGIN)intOrigin, &key_blob)) == EHR_OK) {
+        printf("get key size done 0x%lx\n", key_blob.ulKeyLen);
+        key_blob.pKeyData = (EH_BYTE_PTR)malloc(key_blob.ulKeyLen * sizeof(uint8_t));
+        if (key_blob.pKeyData == NULL) {
+            return "Failed to allow memary\n";
+        }
+
+        rv = CreateKey(intMechanism, (EH_KEY_ORIGIN)intOrigin, &key_blob);
+        if (rv != EHR_OK) {
+            printf("create key 1 failed 0x%lx\n", rv);
+            return "create key 1 failed\n";
+        }
+    } else {
+        printf("create key 2 failed 0x%lx\n", rv);
+        return "create key 2 failed\n";
+    }
+    printf("create key done\n");
+    
+    rest = (char *)malloc(128 * sizeof(uint8_t));
```

Comment:
How to know 128 is big enough? 
Is it possible to use C++ STL and define the base64encoding as:
std::string EncodeBase64(const unsigned char* input_data, int input_num)?

Reply:
FYI:
std::string base64_encoding(const unsigned char* input_data, int input_num) {
const char encode_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
unsigned char batch_input[4] = { 0 };
std::string encode_str;
int length = 0;
for (int i = 0; i < (int)(input_num / 3); i++) {
batch_input[1] = *input_data++;
batch_input[2] = *input_data++;
batch_input[3] = *input_data++;
encode_str += encode_table[batch_input[1] >> 2];
encode_str += encode_table[((batch_input[1] << 4) | (batch_input[2] >> 4)) & 0x3F];
encode_str += encode_table[((batch_input[2] << 2) | (batch_input[3] >> 6)) & 0x3F];
encode_str += encode_table[batch_input[3] & 0x3F];



length +=4;
if (length == 76) {
encode_str += "\r\n";
length = 0;
}
}



int mod = input_num % 3;
if (mod == 2) {
batch_input[1] = *input_data++;
batch_input[2] = *input_data++;
encode_str += encode_table[(batch_input[1] & 0xFC) >> 2];
encode_str += encode_table[((batch_input[1] & 0x03) << 4) | ((batch_input[2] & 0xF0) >> 4)];
encode_str += encode_table[((batch_input[2] & 0x0F) << 2)];
encode_str += "=";
} else {
batch_input[1] = *input_data++;
encode_str += encode_table[(batch_input[1] & 0xFC) >> 2];
encode_str += encode_table[((batch_input[1] & 0x03) << 4)];
encode_str += "==";
}



return encode_str;
}